File: src/main/java/io/github/m/Char.java
package io.github.m;

/**
 * M wrapper class for characters.
 */
public final class Char implements Value.Delegate {
    /**
     * The primitive value of this character.
     */
    public final char value;

    private Char(char value) {
        this.value = value;
    }

    @Override
    public String toString() {
        return Character.toString(value);
    }

    /**
     * Delegates this character to the natural number whose value us the Unicode
     * code point of this character.
     */
    @Override
    public Value value() {
        return Nat.valueOf((int) value);
    }

    /**
     * Wraps a primitive character in a character.
     */
    public static Char valueOf(char value) {
        return new Char(value);
    }

    /**
     * Converts a value to a character.
     */
    public static Char from(Value value) {
        if (value instanceof Char) {
            return (Char) value;
        } else {
            return new Char((char) Nat.from(value).value);
        }
    }

    @MField(name = "space")
    public static final Value space = new Char(' ');

    @MField(name = "tab")
    public static final Value tab = new Char('\t');

    @MField(name = "linefeed")
    public static final Value linefeed = new Char('\n');

    @MField(name = "carriage-return")
    public static final Value carriageReturn = new Char('\r');

    @MField(name = "char.=")
    public static final Value eq = new Value.Impl2("char.=", (x, y) -> Bool.valueOf(from(x).value == from(y).value));

    @MField(name = "char->nat")
    public static final Value toNat = new Value.Impl1("char->nat", value -> Nat.valueOf((int) from(value).value));
}

File: src/main/kotlin/io/github/m/Char.kt
package io.github.m

/**
 * M wrapper class for chars.
 */
@UseExperimental(ExperimentalUnsignedTypes::class)
data class Char(val char: kotlin.Char) : Value.Delegate {
    override fun toString() = char.toString()

    override fun value() = Nat(char.toInt())

    companion object {
        fun from(value: Value) = value as? Char ?: Char(Nat.from(value).value.toChar())
    }

    /**
     * M char definitions.
     */
    @Suppress("unused")
    object Definitions {
        @MField(name = "char.=")
        @JvmField
        val eq: Value = Value.Impl2 { x, y -> Bool(from(x).char == from(y).char) }

        @MField(name = "char->nat")
        @JvmField
        val toNat: Value = Value.Impl1 { x -> Nat(from(x).char.toInt()) }

        @MField(name = "space")
        @JvmField
        val space: Value = Char(' ')

        @MField(name = "tab")
        @JvmField
        val tab: Value = Char('\t')

        @MField(name = "linefeed")
        @JvmField
        val linefeed: Value = Char('\n')

        @MField(name = "carriage-return")
        @JvmField
        val carriageReturn: Value = Char('\r')
    }
}